File: frontend/src/trackpoint.ts
/** Wire layout of a Track Point from /api/tracks:
 * [lon, lat, t_offset_s, hr, pace_s_per_mi] — a compact positional tuple
 * (deck.gl-friendly). This module is the only place that order is known;
 * every consumer goes through these accessors, never by hand-index.
 * Backend mirror: backend/app/trackpoint.py — the two must change together.
 */

export type TrackPoint = [number, number, number, number | null, number | null];

export const lon = (p: TrackPoint): number => p[0];
export const lat = (p: TrackPoint): number => p[1];
export const tOffsetS = (p: TrackPoint): number => p[2];
export const hr = (p: TrackPoint): number | null => p[3];
export const paceSPerMi = (p: TrackPoint): number | null => p[4];

export const metricValue = (p: TrackPoint, metric: "hr" | "pace"): number | null =>
  metric === "hr" ? hr(p) : paceSPerMi(p);

File: backend/app/trackpoint.py
"""Wire layout of a Track Point as served by /api/tracks.

One point is [lon, lat, t_offset_s, hr, pace_s_per_mi] — a compact positional
tuple (deck.gl-friendly). This module is the only place that order is known:
consumers use the named indices, and the SELECT list is derived from the same
definition. The frontend mirror lives in frontend/src/trackpoint.ts; the two
must change together.
"""

WIRE_COLUMNS = ("lon", "lat", "t_offset_s", "hr", "pace_s_per_mi")
WIRE_SELECT = ", ".join(WIRE_COLUMNS)

LON, LAT, T_OFFSET_S, HR, PACE_S_PER_MI = range(len(WIRE_COLUMNS))
